In [1]:
import pandas as pd

Let's load the raw data extracted from https://aqicn.org/. It consist of 40M rows of data!

In [2]:
df = pd.read_csv("delhi_air_quality_feature_store_unprocessed.csv")

In [3]:
df.columns

Index(['location_id', 'city', 'event_timestamp', 'temperature', 'humidity',
       'pressure', 'wind_speed', 'wind_direction', 'pm25', 'pm10', 'no2',
       'so2', 'o3', 'co', 'aqi'],
      dtype='object')

Here AQI will be our target feature for the first use case

In [4]:
df.info()  #this shows that are there are too many missing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4468848 entries, 0 to 4468847
Data columns (total 15 columns):
 #   Column           Dtype  
---  ------           -----  
 0   location_id      object 
 1   city             object 
 2   event_timestamp  object 
 3   temperature      float64
 4   humidity         float64
 5   pressure         float64
 6   wind_speed       float64
 7   wind_direction   float64
 8   pm25             float64
 9   pm10             float64
 10  no2              float64
 11  so2              float64
 12  o3               float64
 13  co               float64
 14  aqi              object 
dtypes: float64(11), object(4)
memory usage: 511.4+ MB


In [5]:
df.describe()

temperature      humidity      pressure    wind_speed  wind_direction  \
count  2.960250e+06  2.960250e+06  2.960250e+06  3.822306e+06    2.744736e+06   
mean   3.000208e+01  5.905234e+01  9.851221e+02  9.269084e-01    1.664309e+02   
std    3.599325e+00  1.707523e+01  7.481772e+00  8.515651e-01    1.062143e+02   
min    2.050000e+01  3.830000e+01  9.647500e+02  1.000000e-01    2.000000e-01   
25%    2.730000e+01  4.406667e+01  9.805000e+02  3.000000e-01    5.840000e+01   
50%    3.030000e+01  5.200000e+01  9.850500e+02  5.333333e-01    1.749000e+02   
75%    3.310000e+01  7.550000e+01  9.891250e+02  1.500000e+00    2.491000e+02   
max    3.575000e+01  9.400000e+01  1.000000e+03  3.600000e+00    3.554000e+02   

               pm25          pm10           no2           so2            o3  \
count  3.822306e+06  3.606792e+06  4.037820e+06  3.175764e+06  4.037820e+06   
mean   2.280420e+02  2.389327e+02  2.472873e+01  6.933108e+00  2.360346e+01   
std    6.139814e+01  1.359065e+02  1.606741e+01  3.582553e+00  2.710938e+01   
min    1.590000e+02  1.090000e+02  3.700000e+00  2.000000e-01  5.000000e-01   
25%    1.900000e+02  1.570000e+02  1.360000e+01  4.000000e+00  4.800000e+00   
50%    2.120000e+02  1.860000e+02  2.080000e+01  6.900000e+00  1.260000e+01   
75%    2.550000e+02  2.710000e+02  3.050000e+01  8.700000e+00  3.250000e+01   
max    4.670000e+02  9.300000e+02  9.080000e+01  2.020000e+01  1.178000e+02   

                 co  
count  4.037820e+06  
mean   1.462703e+01  
std    1.540445e+01  
min    1.000000e-01  
25%    1.000000e-01  
50%    1.120000e+01  
75%    2.400000e+01  
max    6.930000e+01

In [6]:
df.head()

location_id          city  \
0  Delhi Institute of Tool Engineering, Wazirpur,...      Wazirpur   
1             Satyawati College, Delhi, Delhi, India         Delhi   
2  ITI Shahdra, Jhilmil Industrial Area, Delhi, D...       Jhilmil   
3  Sonia Vihar Water Treatment Plant DJB, Delhi, ...   Sonia Vihar   
4   PGDAV College, Sriniwaspuri, Delhi, Delhi, India  Sriniwaspuri   

              event_timestamp  temperature  humidity  pressure  wind_speed  \
0  2000-03-15 13:44:28.651396       30.450    74.100   990.425       0.300   
1  2000-03-15 13:44:28.651396       26.675    74.350   980.650       1.375   
2  2000-03-15 13:44:28.651396       28.775    62.925   984.350       0.300   
3  2000-03-15 13:44:28.651396       26.575    78.525   991.300       0.300   
4  2000-03-15 13:44:28.651396       22.300    71.425   985.650       0.300   

   wind_direction   pm25   pm10   no2   so2    o3    co  aqi  
0            21.7  252.0  332.0  62.4   4.8  22.4  35.7  332  
1            55.3  256.0  234.0  59.2  11.9   2.1  33.8  256  
2           221.3  265.0  210.0  48.3   5.1   8.2  35.1  265  
3           313.6  251.0  177.0  16.5   1.8   1.7  24.9  251  
4           347.1  282.0  347.0  43.1   5.2   3.4  29.9  347

In [7]:
df.isnull().sum()  #to check the null values

location_id              0
city                     0
event_timestamp          0
temperature        1508598
humidity           1508598
pressure           1508598
wind_speed          646542
wind_direction     1724112
pm25                646542
pm10                862056
no2                 431028
so2                1293084
o3                  431028
co                  431028
aqi                      0
dtype: int64

Too many null values in multiple feature columns, will have to fix that.

In [8]:
df.sample(10)  #here got to know that some of the rows have aqi value as "-"

location_id  \
128438   ITI Shahdra, Jhilmil Industrial Area, Delhi, D...   
1199099              ITI Jahangirpuri, Delhi, Delhi, India   
3461524                           ITO, Delhi, Delhi, India   
2284870  Major Dhyan Chand National Stadium, Delhi, Del...   
3032064  Delhi Institute of Tool Engineering, Wazirpur,...   
3856167  Delhi Institute of Tool Engineering, Wazirpur,...   
750015   Delhi Institute of Tool Engineering, Wazirpur,...   
3639567  Delhi Institute of Tool Engineering, Wazirpur,...   
3701139               Burari Crossing, Delhi, Delhi, India   
2445240  Delhi Institute of Tool Engineering, Wazirpur,...   

                                       city             event_timestamp  \
128438                              Jhilmil  2000-11-25 09:44:28.651396   
1199099                        Jahangirpuri  2006-10-10 23:44:28.651396   
3461524                                 ITO  2019-02-06 23:44:28.651396   
2284870  Major Dhyan Chand National Stadium  2012-09-03 07:44:28.651396   
3032064                            Wazirpur  2016-09-24 20:44:28.651396   
3856167                            Wazirpur  2021-05-09 04:44:28.651396   
750015                             Wazirpur  2004-04-11 16:44:28.651396   
3639567                            Wazirpur  2020-02-12 22:44:28.651396   
3701139                              Burari  2020-06-20 04:44:28.651396   
2445240                            Wazirpur  2013-07-18 12:44:28.651396   

         temperature   humidity    pressure  wind_speed  wind_direction  \
128438     28.133333  66.833333  984.366667         0.3           221.3   
1199099    28.550000  59.800000  975.250000         0.3           166.0   
3461524          NaN        NaN         NaN         NaN             NaN   
2284870    31.150000  42.250000  990.150000         0.8           168.0   
3032064    33.700000  50.700000  965.050000         0.3           350.4   
3856167    33.700000  50.700000  965.050000         0.3           350.4   
750015     28.733333  87.066667  991.900000         0.3            23.9   
3639567    33.700000  50.700000  965.050000         0.3           350.4   
3701139          NaN        NaN         NaN         1.5             NaN   
2445240    35.033333  50.633333  973.800000         0.3            14.7   

          pm25   pm10   no2   so2    o3    co  aqi  
128438   265.0  210.0  48.3   5.1   8.2  35.1  265  
1199099  442.0  703.0  30.2   7.5   8.7  24.4  703  
3461524    NaN    NaN   NaN   NaN   NaN   NaN    -  
2284870  235.0  189.0  14.9   7.8  48.8   9.3  235  
3032064  195.0  170.0  25.4  11.5  84.3   2.8  195  
3856167  195.0  170.0  25.4  11.5  84.3   2.8  195  
750015   370.0  766.0  46.9   8.3  28.0  48.6  766  
3639567  195.0  170.0  25.4  11.5  84.3   2.8  195  
3701139  160.0    NaN  17.6   NaN  15.9   0.1  160  
2445240  258.0  325.0  35.2   9.0  60.1   4.9  325

In [9]:
df.shape

(4468848, 15)

In [10]:
(df['aqi'] == '-').sum()

685379

Around 15 percent of the total data have aqi as "-", so we need to take care of it.
But before that let's find out the main sources which are leading to such data, for future preventions and smooth pipeline working.

Let's find the root cause along with dealing with this anomaly

In [11]:
rows_with_no_aqi = df[df['aqi'] == '-']
cities_with_no_aqi_count = rows_with_no_aqi.groupby('city').size().reset_index(name='count')
print(cities_with_no_aqi_count)

           city   count
0   Anand Vihar    3381
1   Civil Lines  215514
2         IHBAS  215514
3           ITO  215514
4   Mandir Marg    3403
5    Parparganj   11692
6          Pusa    3492
7   Sonia Vihar    8275
8  Sriniwaspuri    8594


So multiple cities are giving values as '-'. For now we will just drop all these rows and for future we will make a check that if aqi is '-'. Don't add it to the data.

In [12]:
df = df[df['aqi'] != '-']
df['aqi'] = df['aqi'].astype(float)  # Convert column to numeric type after filtering

In [13]:
(df['aqi'] == '-').sum()

0

In [14]:
df.sample(10)

location_id  \
3195490   PGDAV College, Sriniwaspuri, Delhi, Delhi, India   
920685   Sonia Vihar Water Treatment Plant DJB, Delhi, ...   
3338346                   Anand Vihar, Delhi, Delhi, India   
2742431                   Anand Vihar, Delhi, Delhi, India   
3540751                          Pusa, Delhi, Delhi, India   
2380730  ITI Shahdra, Jhilmil Industrial Area, Delhi, D...   
812706                     R.K. Puram, Delhi, Delhi, India   
4266882                  North Campus, Delhi, Delhi, India   
722974   Major Dhyan Chand National Stadium, Delhi, Del...   
3583627  Delhi Institute of Tool Engineering, Wazirpur,...   

                                       city             event_timestamp  \
3195490                        Sriniwaspuri  2017-08-15 02:44:28.651396   
920685                          Sonia Vihar  2005-04-06 14:44:28.651396   
3338346                         Anand Vihar  2018-05-26 08:44:28.651396   
2742431                         Anand Vihar  2015-02-28 03:44:28.651396   
3540751                                Pusa  2019-07-22 01:44:28.651396   
2380730                             Jhilmil  2013-03-12 12:44:28.651396   
812706                           R.K. Puram  2004-08-14 01:44:28.651396   
4266882                        North Campus  2023-09-11 19:44:28.651396   
722974   Major Dhyan Chand National Stadium  2004-02-18 00:44:28.651396   
3583627                            Wazirpur  2019-10-19 09:44:28.651396   

         temperature   humidity    pressure  wind_speed  wind_direction  \
3195490    29.600000  39.250000  987.600000    0.325000           135.6   
920685     24.680000  85.740000  992.920000    0.300000           313.5   
3338346    31.800000  42.950000  979.450000    0.950000           245.0   
2742431    32.500000  43.600000  979.733333    1.766667           249.1   
3540751    32.150000  44.850000  985.800000    0.300000           180.6   
2380730    34.433333  39.433333  984.066667    0.200000            46.2   
812706     21.950000  87.550000  982.000000    0.300000           175.0   
4266882          NaN        NaN         NaN    3.600000             NaN   
722974     23.700000  80.150000  989.225000    0.475000           229.4   
3583627    33.700000  50.700000  965.050000    0.300000           350.4   

          pm25   pm10   no2   so2     o3    co    aqi  
3195490  192.0  139.0  16.1   5.2  103.0  14.7  192.0  
920685   281.0  179.0  15.2   4.5    1.6  35.8  281.0  
3338346  225.0  179.0  19.8   7.5   19.4   9.5  225.0  
2742431  248.0  218.0  16.0   7.5   14.2  11.3  248.0  
3540751  182.0  116.0  73.1   3.1    1.4  16.0  182.0  
2380730  239.0  198.0  18.0  11.3   16.8  14.8  239.0  
812706   344.0  308.0   9.2  10.6    7.6  35.1  344.0  
4266882  164.0  157.0   7.7   NaN   12.6   0.1  164.0  
722974   278.0  180.0  40.3   7.7    8.4  36.6  278.0  
3583627  195.0  170.0  25.4  11.5   84.3   2.8  195.0

In [15]:
df.isnull().sum()

location_id              0
city                     0
event_timestamp          0
temperature        1077570
humidity           1077570
pressure           1077570
wind_speed          215514
wind_direction     1077570
pm25                     0
pm10                215514
no2                      0
so2                 862056
o3                       0
co                       0
aqi                      0
dtype: int64

Let's handle the rows which have missing pm10 values

Let's find out the number of rows which do not have pm10 and of the other anomalies either

In [16]:
rows_without_data = df[
    df['so2'].isna() &
    df['pm10'].isna() &
    df['temperature'].isna() &
    df['humidity'].isna() &
    df['pressure'].isna()
]

count_rows_without_data = len(rows_without_data)

print(f'Number of rows without so2, pm10, temperature, humidity, and pressure data: {count_rows_without_data}')

Number of rows without so2, pm10, temperature, humidity, and pressure data: 215514


So the common rows with all the anomalies are: 215514

In [17]:
# 1. Find rows where pm10 is missing
rows_without_pm10 = df[df['pm10'].isna()]

# 2. Find rows where pm10 is missing and also so2, temperature, humidity, and pressure are missing
rows_without_all = rows_without_pm10[
    rows_without_pm10['so2'].isna() &
    rows_without_pm10['temperature'].isna() &
    rows_without_pm10['humidity'].isna() &
    rows_without_pm10['pressure'].isna()
]

# Count the number of rows without pm10
count_rows_without_pm10 = len(rows_without_pm10)

# Count the number of rows without pm10 and the other columns
count_rows_without_all = len(rows_without_all)

print(f'Number of rows without pm10: {count_rows_without_pm10}')
print(f'Out of these, number of rows also missing so2, temperature, humidity, and pressure: {count_rows_without_all}')


#this concludes that all the rows which do not have pm10 simply do not have any of them, so we will first drop all these rows which have do not have pm10

Number of rows without pm10: 215514
Out of these, number of rows also missing so2, temperature, humidity, and pressure: 215514


This shows that the rows which have missing pm10 value, do not have any of the other values either!

Let's find out which city has provided how many such rows without pm10?

In [18]:
rows_without_pm10 = df[df['pm10'].isna()]

rows_without_all = rows_without_pm10[
    rows_without_pm10['so2'].isna() &
    rows_without_pm10['temperature'].isna() &
    rows_without_pm10['humidity'].isna() &
    rows_without_pm10['pressure'].isna()
]

#Group by city and count the number of rows without pm10
count_by_city_pm10 = rows_without_pm10.groupby('city').size().reset_index(name='count_without_pm10')

#Group by city and count the number of rows without pm10 and the other specified columns
count_by_city_all = rows_without_all.groupby('city').size().reset_index(name='count_without_all')

# Merge both counts to get a combined view
merged_count = pd.merge(count_by_city_pm10, count_by_city_all, on='city', how='left')

# Display the result
print(merged_count)

#this shows that burari is the impostor:) all these without pm10 val data are provided by burari only

     city  count_without_pm10  count_without_all
0  Burari              215514             215514


Total rows with burari as city

That's concludes that all the rows which have missing so2 values along with other anomalies are coming from burari!. So we will remove the rows from this city and refain it from future repo too

In [19]:
# Filter the rows where the city is 'Burari'
rows_burari = df[df['city'] == 'Burari']

# Count the number of rows for Burari
count_burari_rows = len(rows_burari)

print(f'Total number of rows for Burari: {count_burari_rows}')

#total rows from burari are 215514 and all of them do not have any pm10 values

Total number of rows for Burari: 215514


Conclusion: Only the rburari data lacks pm10 along with other values, so simply remove burari

In [20]:
#so we will remove all the data of burari
df = df[df['city'] != 'Burari']

In [21]:
df.isnull().sum()

location_id             0
city                    0
event_timestamp         0
temperature        862056
humidity           862056
pressure           862056
wind_speed         215514
wind_direction     862056
pm25                    0
pm10                    0
no2                     0
so2                646542
o3                      0
co                      0
aqi                     0
dtype: int64

Now lets investigate so2

The Mathura road provides data which is do not contain temp, humidity, wind speed, wind direction and so2. So it's of no use to us.

In [22]:
df = df[df['city'] != 'Mathura Road']

Let's identify the rows that are missing so2 values and determine how many of these rows are also missing other features.

In [23]:
# 1. Find rows where so2 is missing
rows_without_pm10 = df[df['so2'].isna()]

# 2. Find rows where pm10 is missing and also so2, temperature, humidity, and pressure are missing
rows_without_all = rows_without_pm10[
    rows_without_pm10['temperature'].isna() &
    rows_without_pm10['humidity'].isna() &
    rows_without_pm10['pressure'].isna()
]

# Count the number of rows without pm10
count_rows_without_pm10 = len(rows_without_pm10)

# Count the number of rows without pm10 and the other columns
count_rows_without_all = len(rows_without_all)

print(f'Number of rows without so2: {count_rows_without_pm10}')
print(f'Out of these, number of rows also missing temperature, humidity, and pressure: {count_rows_without_all}')

Number of rows without so2: 431028
Out of these, number of rows also missing temperature, humidity, and pressure: 431028


Which city has provided how many such rows without so2?

In [24]:
# 1. Filter rows where so2 is missing (NaN)
rows_without_so2 = df[df['so2'].isna()]

# 2. Filter rows where so2 is missing and temperature, humidity, and pressure are also missing
rows_without_all = rows_without_so2[
    rows_without_so2['temperature'].isna() &
    rows_without_so2['humidity'].isna() &
    rows_without_so2['pressure'].isna()
]

# 3. Group by city and count the number of rows without so2
count_by_city_so2 = rows_without_so2.groupby('city').size().reset_index(name='count_without_so2')

# 4. Group by city and count the number of rows without so2 and the other specified columns
count_by_city_all = rows_without_all.groupby('city').size().reset_index(name='count_without_all')

# Merge both counts to get a combined view (using outer join to handle missing cities)
merged_count = pd.merge(count_by_city_so2, count_by_city_all, on='city', how='outer')

# Display the result
print(merged_count)

#this shows that burari is the impostor:) all these without pm10 val data are provided by burari only

           city  count_without_so2  count_without_all
0    Lodhi Road             215514             215514
1  North Campus             215514             215514


#Now here we can see that:
Total rows without so2 value = 431028
Out of which :
1. Lodhi Road Contributed: 215514
2. North Campus Contributed: 215514
So both are 50-50 percent impostor

Now lets see the number of total rows which have cities as lodhi road and north campus


In [25]:
# Filter the rows where the city is 'Lodhi Road' and North Campus
rows_lodhi_road = df[df['city'] == 'Lodhi Road']
rows_north_campus = df[df['city'] == 'North Campus']


# Count the number of rows for Lodhi Road
count_lodhi_road_rows = len(rows_lodhi_road)
count_north_campus_rows = len(rows_north_campus)

print(f'Total number of rows for Lodhi Road: {count_lodhi_road_rows}')
print(f'Total number of rows for North Campus: {count_north_campus_rows}')

Total number of rows for Lodhi Road: 215514
Total number of rows for North Campus: 215514



This isn't ideal, as both cities lack so2 data, which could negatively impact the accuracy of any analysis. Before removing the rows, it's important to understand the role of so2 in the Air Quality Index (AQI) and its characteristics:

Impact on AQI: Sulfur dioxide (SO₂) is a critical pollutant used to calculate AQI. High concentrations of SO₂ can lead to respiratory issues and contribute to the formation of other pollutants like particulate matter and acid rain. Its levels play a significant role in assessing overall air quality.
Properties: SO₂ levels can differ greatly between locations due to factors like industrial emissions, vehicle pollution, and climatic conditions. Even cities in close proximity may show significant differences in SO₂ concentration because of these factors.
Given that no so2 data is available for these two cities, the best course of action is to drop these rows to maintain the integrity of the dataset.

In [26]:
df = df[df['city'] != 'Lodhi Road']
df = df[df['city'] != 'North Campus']

In [27]:
df.shape

(2921413, 15)

In [28]:
df.isna().sum()

location_id             0
city                    0
event_timestamp         0
temperature        215514
humidity           215514
pressure           215514
wind_speed         215514
wind_direction     215514
pm25                    0
pm10                    0
no2                     0
so2                     0
o3                      0
co                      0
aqi                     0
dtype: int64

So there are still some rows which have temperature,humidity,	pressure, wind_speed, wind_direction. Lets find from which cities they are.

In [29]:
filtered_cities = df[(df[['temperature', 'humidity', 'pressure', 'wind_speed', 'wind_direction']].isnull().any(axis=1))]

cities_missing_weather_data = filtered_cities['city'].unique()

print("Cities with missing temperature, humidity, pressure, wind speed, or wind direction:")
print(cities_missing_weather_data)

Cities with missing temperature, humidity, pressure, wind speed, or wind direction:
['Siri Fort']


Siri Fort is now the only one remaining anomaly

In [30]:
df.shape

(2921413, 15)

Siri Fort does not provide values for temperature, humidity, wind speed, or wind direction but does provide other important factors. Therefore, we will update these missing values for each hour of the day with the average values from all the other cities for that same hour.

In [31]:
# Convert event_timestamp to datetime format if not already done
df['event_timestamp'] = pd.to_datetime(df['event_timestamp'])

# Extract hour-based grouping key
df['hour_key'] = df['event_timestamp'].dt.strftime('%Y-%m-%d %H')

# Calculate average values for each hour, excluding Siri Fort
hourly_avg = df[df['city'] != 'Siri Fort'].groupby('hour_key')[
    ['temperature', 'humidity', 'wind_speed', 'wind_direction', 'pressure']
].mean()

# Merge the averages with Siri Fort rows
siri_fort_indices = df[df['city'] == 'Siri Fort'].index
df.loc[siri_fort_indices, ['temperature', 'humidity', 'wind_speed', 'wind_direction', 'pressure']] = (
    df.loc[siri_fort_indices, 'hour_key']
    .map(hourly_avg.to_dict(orient='index'))
    .apply(pd.Series)
)

# Round the numerical columns to 3 decimal places
df[['temperature', 'humidity', 'wind_speed', 'wind_direction', 'pressure']] = df[
    ['temperature', 'humidity', 'wind_speed', 'wind_direction', 'pressure']
].round(3)

# Drop the temporary 'hour_key' column
df.drop(columns=['hour_key'], inplace=True)


In [32]:
df.isna().sum()

location_id        0
city               0
event_timestamp    0
temperature        0
humidity           0
pressure           0
wind_speed         0
wind_direction     0
pm25               0
pm10               0
no2                0
so2                0
o3                 0
co                 0
aqi                0
dtype: int64

In [33]:
df.head()

location_id          city  \
0  Delhi Institute of Tool Engineering, Wazirpur,...      Wazirpur   
1             Satyawati College, Delhi, Delhi, India         Delhi   
2  ITI Shahdra, Jhilmil Industrial Area, Delhi, D...       Jhilmil   
3  Sonia Vihar Water Treatment Plant DJB, Delhi, ...   Sonia Vihar   
4   PGDAV College, Sriniwaspuri, Delhi, Delhi, India  Sriniwaspuri   

             event_timestamp  temperature  humidity  pressure  wind_speed  \
0 2000-03-15 13:44:28.651396       30.450    74.100   990.425       0.300   
1 2000-03-15 13:44:28.651396       26.675    74.350   980.650       1.375   
2 2000-03-15 13:44:28.651396       28.775    62.925   984.350       0.300   
3 2000-03-15 13:44:28.651396       26.575    78.525   991.300       0.300   
4 2000-03-15 13:44:28.651396       22.300    71.425   985.650       0.300   

   wind_direction   pm25   pm10   no2   so2    o3    co    aqi  
0            21.7  252.0  332.0  62.4   4.8  22.4  35.7  332.0  
1            55.3  256.0  234.0  59.2  11.9   2.1  33.8  256.0  
2           221.3  265.0  210.0  48.3   5.1   8.2  35.1  265.0  
3           313.6  251.0  177.0  16.5   1.8   1.7  24.9  251.0  
4           347.1  282.0  347.0  43.1   5.2   3.4  29.9  347.0

In [34]:
df.to_csv("delhi_air_quality_feature_store_processed.csv", index=False)

From next we will make sure that we don't get the data from burari, north campus and lodhi road